betterproto/plugin.py

```diff
@@ -119,265 +130,547 @@ def get_comment(proto_file, path: List[int], indent: int = 4) -> str:
     return ""
 
 
-def generate_code(request, response):
-    plugin_options = request.parameter.split(",") if request.parameter else []
+class ProtoContentBase:
+    """Methods common to Message, Service and ServiceMethod."""
+
+    path: List[int]
+    comment_indent: int = 4
+
+
+    def __post_init__(self):
+        """Checks that no fake default fields were left as placeholders."""
+        for field_name, field_val in self.__dataclass_fields__.items():
+            if field_val is PLACEHOLDER:
+                raise ValueError(f"`{field_name}` is a required field with no default value.")
+
+    @property   # TODO: replace with functools.cached_property
```

Comment:
This would bump the required version to generate protobufs, might be better to just copy the source for it somewhere

Reply:
Yeah. I see three alternatives:

1. Copy it from source (as you suggested).
2. Leave it as `@property` which means it will be recalculated every time.
3. Use `@functools.lru_cache(maxsize=1)` which would achieve the same thing, albeit with overkill complexity in the backend.

Reply:
I tested pulling the source, it's pretty straightforward. Only question is if it's worth the trouble or if the performance is okay as is.